Expose Dyson PureCool filter life remaining percentage

Flameeyes opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add sensors for "life remaining percentage" for Dyson PureCool and PureCool+Humidify filters.

These are different from the life in hours that is exposed for other devices.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

https://github.com/home-assistant/home-assistant.io/pull/15563

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Flameeyes committed:
Dyson: expose the filters' life remaining percentage for PureCool.

PureCool (and PureCool+Humidify) have filters that exposes a life
remaining percentage, similarly to a printer's drum.

Until https://github.com/etheralm/libpurecool/issues/16 is merged, the
PureCool+Humidify objects are not "aware" of the Combi filter directly,
and instead expose its life as if it was the HEPA filter.

homeassistant commented:
Hi @Flameeyes, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42765).

Once you do that we will be able to review and accept this pull request.

Thanks!

homeassistant commented:
Hi @Flameeyes, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42765).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please sign the CLA and extend the tests to reach full coverage.

Flameeyes commented:
> Please sign the CLA and extend the tests to reach full coverage.

Yep, currently waiting for approval by my employer, as the CLA requires manual review.

If/once I get the approval I'll also send a pull request to document the presence of the CLA as part of the contributing guidelines, as that took me by surprise.

homeassistant commented:
Hi @Flameeyes, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42765).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare commented:
For the future, please don't squash commits after review has started to make it easier for readers to track changes. Thanks!

MartinHjelmare commented on tests/components/dyson/test_sensor.py:
```diff
@@ -137,6 +190,83 @@ def test_dyson_filter_life_sensor_with_values(self):
         assert sensor.entity_id == "sensor.dyson_1"
         sensor.on_message("message")
 
+    def test_dyson_hepa_filter_state_sensor(self):
```
For the new tests, please instead write the tests like this as standalone async test functions and use the core interfaces:
https://github.com/home-assistant/core/blob/aecc3476e84fcdd87675c381c9eea01247f2a338/tests/components/dyson/test_climate.py#L213-L244

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Flameeyes replied:
Hah, it turns out it's actually easier to test it this way, and it tests the one Combi selection logic trick too. Thanks!

Flameeyes committed:
dyson: use the async standalone test.

This replaces the without-state/with-state tests with two more
streamlined tests that should cover the same interface, while being
easier to read.

In particular this actually tests the logic of selecting the Combi
filter life sensor rather than the HEPA and Carbon separate sensors.

Flameeyes commented:
> For the future, please don't squash commits after review has started to make it easier for readers to track changes. Thanks!

Will do! Sorry, used to a different "rebase and merge" kind of workflow, so it became second nature. This time I added the commit on top.

MartinHjelmare commented on tests/components/dyson/test_sensor.py:
```diff
@@ -403,3 +307,49 @@ async def test_purecool_component_setup_only_once(devices, login, hass):
+    config = _get_config()
+    await async_setup_component(hass, dyson_parent.DOMAIN, config)
+    await hass.async_block_till_done()
+    discovery.load_platform(hass, "sensor", dyson_parent.DOMAIN, {}, config)
```
This isn't needed. The Dyson sensor platform will be loaded when the Dyson integration is set up.

Flameeyes committed:
Remove unneeed load_platform call.

MartinHjelmare commented on tests/components/dyson/test_sensor.py:
```diff
@@ -269,7 +303,47 @@ async def test_purecool_component_setup_only_once(devices, login, hass):
     await hass.async_block_till_done()
-    discovery.load_platform(hass, "sensor", dyson_parent.DOMAIN, {}, config)
+
+    assert len(hass.data[dyson.DYSON_SENSOR_DEVICES]) == 4
```
Assert the number of states in the state machine under the sensor domain instead.

Flameeyes replied:
Do you mind if I back out of changing this test altogether, and just change the number to 4 so that that particular test passes?

I'm not quite sure I understand how this part of the test is meant to work, or what it's exactly meant to test :/

MartinHjelmare replied:
Yeah, that's ok.

MartinHjelmare commented on tests/components/dyson/test_sensor.py:
```diff
@@ -269,7 +303,47 @@ async def test_purecool_component_setup_only_once(devices, login, hass):
     config = _get_config()
     await async_setup_component(hass, dyson_parent.DOMAIN, config)
     await hass.async_block_till_done()
-    discovery.load_platform(hass, "sensor", dyson_parent.DOMAIN, {}, config)
```
This call was probably an artificial way of forcing another platform set up to verify that duplicate entities is not created. Not sure if we really need to test that, since it can't happen in the code under test.

Flameeyes replied:
Restored, just to avoid making more changes that I meant to.

Flameeyes committed:
Back out of changing test_purecool_component_setup_only_once().

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please open a docs PR and update the sensor section. Link the docs PR in this PR description.
https://www.home-assistant.io/integrations/dyson/#sensor

MartinHjelmare commented:
Just two lines missing from coverage. Merging.
